src/librustdoc/html/markdown.rs

```diff
@@ -340,29 +340,86 @@ impl<'a, I: Iterator<Item = Event<'a>>> Iterator for CodeBlocks<'_, 'a, I> {
+        let mut event = self.inner.next();
+
+        // Remove disambiguators from shortcut links (`[fn@f]`)
+        match &mut event {
+            Some(Event::Start(Tag::Link(
+                pulldown_cmark::LinkType::ShortcutUnknown,
+                dest,
+                title,
+            ))) => {
+                debug!("saw start of shortcut link to {} with title {}", dest, title);
+                let link = if let Some(link) =
+                    self.links.iter().find(|&link| *link.original_text == **dest)
+                {
+                    // Not sure why this is necessary - maybe the broken_link_callback doesn't always work?
+                    *dest = CowStr::Borrowed(link.href.as_ref());
+                    Some(link)
+                } else {
+                    self.links.iter().find(|&link| *link.href == **dest)
+                };
+                if let Some(link) = link {
+                    trace!("it matched");
+                    assert!(self.shortcut_link.is_none(), "shortcut links cannot be nested");
+                    self.shortcut_link = Some(link);
+                }
             }
-        } else {
-            event
+            Some(Event::End(Tag::Link(pulldown_cmark::LinkType::ShortcutUnknown, dest, _))) => {
+                debug!("saw end of shortcut link to {}", dest);
+                if let Some(_link) = self.links.iter().find(|&link| *link.href == **dest) {
+                    assert!(self.shortcut_link.is_some(), "saw closing link without opening tag");
+                    self.shortcut_link = None;
+                }
+            }
+            // Handle backticks in inline code blocks
+            Some(Event::Code(text)) => {
+                trace!("saw code {}", text);
+                if let Some(link) = self.shortcut_link {
+                    trace!("original text was {}", link.original_text);
+                    if **text == link.original_text[1..link.original_text.len() - 1] {
+                        debug!("replacing {} with {}", text, link.new_text);
+                        *text = link.new_text.clone().into();
+                    }
+                }
+            }
+            // Replace plain text in links
+            Some(Event::Text(text)) => {
+                trace!("saw text {}", text);
+                if let Some(link) = self.shortcut_link {
+                    trace!("original text was {}", link.original_text);
+                    if **text == *link.original_text {
+                        debug!("replacing {} with {}", text, link.new_text);
+                        *text = link.new_text.clone().into();
+                    }
+                }
+            }
+            Some(Event::Start(Tag::Link(_, dest, _))) => {
+                if let Some(link) = self.links.iter().find(|&link| *link.original_text == **dest) {
+                    // Not sure why this is necessary - maybe the broken_link_callback doesn't always work?
```

Comment:
broken_link_callback only does broken links, not links which have a URL. `[foo](Foo)` links have a URL, even if that URL doesn't resolve to anything